File: algorithms/160.Intersection_of_Two_Linked_Lists/Intersection_of_Two_Linked_Lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param two ListNodes
    # @return the intersected ListNode
    def getIntersectionNode(self, headA, headB):
        curA, curB = headA, headB
        begin, tailA, tailB = None, None, None
        
        while curA and curB:
            if curA == curB:
                begin = curA
                break
                
            if curA.next:
                curA = curA.next
            elif tailA is None:
                tailA = curA
                curA = headB
            else:
                break
            
            if curB.next:
                curB = curB.next
            elif tailB is None:
                tailB = curB
                curB = headA
            else:
                break
        
        return begin

File: algorithms/160.Intersection_of_Two_Linked_Lists/Intersection_of_Two_Linked_Lists.cpp
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int alen = 0, blen = 0;
        ListNode *tmpA = headA, *tmpB = headB;

        // get the length of A, B
        while (tmpA != NULL) {
            alen++;
            tmpA = tmpA->next;
        }
        while (tmpB != NULL) {
            blen++;
            tmpB = tmpB->next;
        }
        if (tmpA || tmpB)
            return NULL;

        // set A be the longer one
        if (alen < blen) {
            swap(headA, headB);
            swap(alen, blen);
        }

        tmpA = headA;
        tmpB = headB;
        // cut A
        while (alen > blen) {
            alen--;
            tmpA = tmpA->next;
        }
        while (blen--) {
            if (tmpA == tmpB)
                return tmpA;
            tmpA = tmpA->next;
            tmpB = tmpB->next;
        }
        return NULL;
    }
};
class Solution {
    public:
        ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {

            //caculate the length of each List
            int lenA = getListLength(headA);
            int lenB = getListLength(headB);

            if (lenA<=0 || lenB<=0 ) return NULL;

            //let List A is the longest List;
            if (lenA < lenB){
                swap(headA, headB);
            }

            //move head of List A, make both of Lists are same length 
            for (int i=0; i<abs(lenA-lenB); i++){
                headA = headA->next;
            }

            //synced travel both of Lists and check their nodes are same or not 
            while (headA != headB){
                headA = headA->next;
                headB = headB->next;
            }

            return headA;
        }
    private:    
        inline int getListLength(ListNode *head){
            int len=0;
            while(head!=NULL){
                head = head->next;
                len++;
            }
            return len;
        }
};